### DISCUSSION TITLE
SQLAlchemy 2.0 models declared with DeclarativeBase not created (on sqlite) by Alembic

### DISCUSSION THREAD

**Discussion by alessandromannini** [sqlalchemy/alembic#1189] [Usage Questions]
Hi,

i defined some model (see attached txt) using the new declarative syntax of SQLAlchemy and tried to generate the first migration using Alembic (using Flask-Migrate) but I receive the message "No changes detected".

I've tried many solutions without success.
As suggested in the comment I also tried to alter env.py in this way:
`# add your model's MetaData object here`
`# for 'autogenerate' support`
`# from myapp import mymodel`
`# target_metadata = mymodel.Base.metadata`
`config.set_main_option('sqlalchemy.url', current_app.config.get('SQLALCHEMY_DATABASE_URI'))`
`target_db = current_app.extensions['migrate'].db`
**`from application import models`**
**`target_metadata = models.Base.metadata`**

If I put an "old style" (with Column) declaration in my models a create migration is generated.
Where i'm worng?

Thx
Alessandro
[models.txt](https://github.com/sqlalchemy/sqlalchemy/files/10844016/models.txt)

**Comment by zzzeek** [comment]
I would try printing out `target_metadata.tables.values()` to see that it's not empty

**Reply by alessandromannini** [reply]
Hi! And thx!

I've already tried (but not written here :D).
It's not empty. This is the output:

```
dict_values([Table('users', MetaData(), Column('id', Integer(), table=<users>, primary_key=True, nullable=False), Column('username', String(), table=<users>, nullable=False), Column('password', String(length=50), table=<users>, nullable=False), Column('created_at', DateTime(), table=<users>, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.functions.current_timestamp at 0x1f16d941c10; current_timestamp>, for_update=False)), Column('modified_at', DateTime(), table=<users>, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.functions.current_timestamp at 0x1f16bd17b90; current_timestamp>, for_update=False)), schema=None), Table('layouts', MetaData(), Column('id', Integer(), table=<layouts>, primary_key=True, nullable=False), Column('name', String(length=50), table=<layouts>, nullable=False), Column('gridRows', Integer(), table=<layouts>, nullable=False, default=ScalarElementColumnDefault(0)), Column('gridCols', Integer(), table=<layouts>, nullable=False, default=ScalarElementColumnDefault(0)), Column('created_at', DateTime(), table=<layouts>, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.functions.current_timestamp at 0x1f16d956810; current_timestamp>, for_update=False)), Column('modified_at', DateTime(), table=<layouts>, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.functions.current_timestamp at 0x1f16d956a50; current_timestamp>, for_update=False)), schema=None), Table('nodes', MetaData(), Column('id', Integer(), table=<nodes>, primary_key=True, nullable=False), Column('name', String(length=50), table=<nodes>, nullable=False), Column('type', String(length=2), table=<nodes>, nullable=False), Column('layout_id', Integer(), ForeignKey('layout.id'), table=<nodes>), Column('gridRow', Integer(), table=<nodes>, nullable=False, default=ScalarElementColumnDefault(-1)), Column('gridCol', Integer(), table=<nodes>, nullable=False, default=ScalarElementColumnDefault(-1)), Column('orientation', String(length=2), table=<nodes>, nullable=False), Column('created_at', DateTime(), table=<nodes>, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.functions.current_timestamp at 0x1f16d961650; current_timestamp>, for_update=False)), Column('modified_at', DateTime(), table=<nodes>, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.functions.current_timestamp at 0x1f16d961890; current_timestamp>, for_update=False)), schema=None)])
```


But i noticed that metadata are not taken from the suggested var target_metadata but from the function get_metadata() 😮 that look to the target_db.... . Let see the code:

```def get_metadata():
    if hasattr(target_db, 'metadatas'):
        return target_db.metadatas[None]
    return target_db.metadata`

...[CUT]...

   with connectable.connect() as connection:
        context.configure(
            connection=connection,
            target_metadata=get_metadata(),
            process_revision_directives=process_revision_directives,
            **current_app.extensions['migrate'].configure_args
        )

        with context.begin_transaction():
            context.run_migrations()`
```

In effect using target_metadata in context.configure... the (first) migration is created!!

**Reply by CaselIT** [reply]
Hi, what is `get_metadata` or `target_db`?

This does not seem the standard alembic tamplate. Are using a version that's created by flask migrate?
Please ensure that the correct metadata is provided to alembic. It appears that flask migrate does something that make it not work by default with the v2 style.

Maybe you need to tell the Base to use the same metadata used by flask migrate? You can do that by doing
```py
class Base(DeclarativeBase):
   ...
   metadata = the_flask_migrate_metadata_here
```

**Reply by alessandromannini** [reply]
Hi, if you look at the code in my previous posts:
- target_db = current_app.extensions['migrate'].db
- get_metadata() is the function to get metadata from target_db

...but you are right I'm using the env.py generated by flask-migrate that I supposed to be alembic-standard but...I understand...it isn't.

So it's better to move the question to flask-migrate.

Thanks
A

**Reply by CaselIT** [reply]
I think the most likely solution is that you have to make base use the same metadata of flask-migrate. Alternatively you have to tell flask migrate to use the other metadata (don't know if it supports that)

**Reply by miguelgrinberg** [reply]
Just so that we are all on the same page, please keep in mind that Flask-Migrate requires the Flask-SQLAlchemy or Alchemical extensions for Flask, and it provides its own customized template that integrate with these extensions. To work directly with SQLAlchemy as the OP is intending, the env.py file would need to be changed and made to look more like the official Alembic template, but I would suggest using Alembic directly in that case.

**Reply by CaselIT** [reply]
I imagined something similar was the case. You should be able to share the same metadata with that's used by flaks-sqlalchemy by setting it in the Base as mentioned above. that should allow inter-operation without much trouble, since the changes for v2 are mostly "cosmetic"

**Comment by CaselIT** [comment]
(moved to alembic since this is not a sqlalchemy issue)
